File: src/ssh_manager/services/importer.py
"""Onboard an existing ~/.ssh into the manifest + inventory.

`import` parses an existing ssh config (following ``Include`` directives) into
manifest profiles + hosts, and fingerprints any private keys it can find into the
inventory - so you never hand-write JSON. Profile assignment derives from an
``IdentityFile`` path under ``profiles/<profile>/`` when present, else ``imported``.
"""
from __future__ import annotations

import contextlib
import glob
import os
import shutil
from dataclasses import dataclass, field
from pathlib import Path

from ..core.inventory import Inventory, KeyRecord
from ..core.key import derive_key_name
from ..core.manifest import (
    _DANGEROUS_OPTIONS,
    DEFAULT_GLOBAL_OPTIONS,
    Defaults,
    Host,
    Manifest,
    Profile,
)
from ..platforms.base import Platform
from ..util import perms
from ..util.errors import ManifestError
from ..util.paths import Paths
from .keystore import KeyStore


@dataclass
class ParsedHost:
    alias: str
    hostname: str = ""
    user: str = ""
    port: int = 22
    identity_file: str | None = None
    profile: str = "imported"
    extra: dict[str, str] = field(default_factory=dict)


@dataclass
class ImportResult:
    profiles: dict[str, int] = field(default_factory=dict)   # profile -> host count
    keys_found: int = 0
    adopted: int = 0          # existing keys copied into the profiles/ layout
    dry_run: bool = False

    def format(self) -> str:
        head = "import (dry-run):" if self.dry_run else "import:"
        lines = [head]
        for p, n in sorted(self.profiles.items()):
            lines.append(f"  profile {p}: {n} host(s)")
        lines.append(f"  keys fingerprinted into inventory: {self.keys_found}")
        if self.adopted:
            lines.append(f"  keys adopted into profiles/ layout: {self.adopted}")
        return "\n".join(lines)


_SIMPLE_KEYS = {"hostname", "user", "port", "identityfile"}
_PROVIDER_BY_HOST = {
    "github.com": "github", "gitlab.com": "gitlab", "bitbucket.org": "bitbucket",
}


def _infer_provider(hostname: str) -> str | None:
    return _PROVIDER_BY_HOST.get(hostname.lower())


def parse_ssh_config(text: str, *, base_dir: Path | None = None,
                     _seen: set[Path] | None = None) -> list[ParsedHost]:
    """Parse ssh config text into hosts, following relative ``Include`` files.

    Pure over its inputs (``base_dir`` only used to resolve includes); ``Host *``
    and other wildcard/global blocks are skipped - they carry no concrete host. A
    ``Host`` line with several patterns applies its options to *all* of them (as ssh
    does); a ``Match`` line ends the current block (ssh-manager doesn't model Match).
    """
    hosts: list[ParsedHost] = []
    current: list[ParsedHost] = []
    for raw in text.splitlines():
        line = raw.strip()
        if not line or line.startswith("#"):
            continue
        parts = line.split(None, 1)
        keyword = parts[0].lower()
        value = parts[1].strip() if len(parts) > 1 else ""
        if keyword == "host":
            current = []
            for alias in value.split():
                if any(c in alias for c in "*?!"):
                    continue
                ph = ParsedHost(alias=alias)
                current.append(ph)
                hosts.append(ph)
        elif keyword == "match":
            current = []                 # Match block: unmodeled, ends the host block
        elif keyword == "include" and base_dir is not None:
            hosts.extend(_parse_includes(value, base_dir, _seen or set()))
            current = []
        else:
            for ph in current:
                _apply_option(ph, keyword, value)
    return hosts


def _apply_option(host: ParsedHost, keyword: str, value: str) -> None:
    if keyword == "hostname":
        host.hostname = value
    elif keyword == "user":
        host.user = value
    elif keyword == "port":
        with contextlib.suppress(ValueError):
            host.port = int(value)
    elif keyword == "identityfile":
        host.identity_file = value
        host.profile = _profile_from_identity(value)
    elif keyword in _DANGEROUS_OPTIONS:
        # Code/config-executing directives (ProxyCommand, Match, Include, ...) are
        # dropped on import rather than carried into the manifest (where validation
        # would reject them and abort the whole import).
        return
    elif keyword not in _SIMPLE_KEYS:
        host.extra[keyword] = value


def _profile_from_identity(identity_file: str) -> str:
    parts = Path(identity_file).parts
    if "profiles" in parts:
        i = parts.index("profiles")
        if i + 1 < len(parts):
            return parts[i + 1]
    return "imported"


def _parse_includes(pattern: str, base_dir: Path, seen: set[Path]) -> list[ParsedHost]:
    hosts: list[ParsedHost] = []
    for token in pattern.split():
        expanded = os.path.expanduser(token)
        full = expanded if os.path.isabs(expanded) else str(base_dir / expanded)
        for match in sorted(glob.glob(full)):
            mpath = Path(match).resolve()
            if mpath in seen or not mpath.is_file():
                continue
            seen.add(mpath)
            hosts.extend(parse_ssh_config(
                mpath.read_text(encoding="utf-8"), base_dir=base_dir, _seen=seen
            ))
    return hosts


@dataclass
class _Resolution:
    """How one parsed host maps onto the canonical layout."""

    key_name: str
    adopt_from: Path | None = None   # source key to copy into profiles/<p>/ (if any)
    probe: Path | None = None        # existing key to fingerprint (source-side)


class Importer:
    def __init__(self, platform: Platform, paths: Paths) -> None:
        self._platform = platform
        self._paths = paths
        self._keystore = KeyStore(platform)

    def run(self, config_path: Path, *, dry_run: bool = False) -> ImportResult:
        config_path = config_path.expanduser()
        if not config_path.is_file():
            raise ManifestError(f"no ssh config file to import: {config_path}")
        try:
            text = config_path.read_text(encoding="utf-8")
        except OSError as exc:
            raise ManifestError(f"cannot read {config_path}: {exc}") from exc
        parsed = parse_ssh_config(text, base_dir=config_path.parent)
        opts = dict(DEFAULT_GLOBAL_OPTIONS)
        if not self._platform.emits_use_keychain:
            opts.pop("UseKeychain", None)

        # Pass 1: build the full profile map, then the manifest (pydantic copies
        # the dict on construct, so it must be complete before we build it).
        profiles: dict[str, Profile] = {}
        resolved: list[tuple[ParsedHost, _Resolution]] = []
        seen: set[tuple[str, str]] = set()
        for h in parsed:
            if (h.profile, h.alias) in seen:
                continue                 # duplicate Host block in the source - first wins
            seen.add((h.profile, h.alias))
            res = self._resolve_key(h)
            resolved.append((h, res))
            host = Host(
                alias=h.alias, hostname=h.hostname or h.alias,
                user=h.user or os.environ.get("USER", "git"),
                port=h.port, key_name=res.key_name,
                provider=_infer_provider(h.hostname or h.alias),
                raw_options=dict(h.extra),   # carry unmodeled options (ProxyJump, ...)
            )
            profiles.setdefault(h.profile, Profile()).hosts.append(host)
        manifest = Manifest(defaults=Defaults(global_options=opts), profiles=profiles)

        # Pass 2: adopt keys into the profiles/ layout + fingerprint into inventory.
        # The recorded path is the canonical IdentityFile the renderer will emit,
        # so import + reconcile never disagree (no phantom-key minting).
        inventory = Inventory()
        adopted = 0
        for h, res in resolved:
            ident = manifest.identity_file(h.profile, res.key_name)
            if (not dry_run and res.adopt_from is not None
                    and self._adopt_key(res.adopt_from, h.profile, res.key_name)):
                adopted += 1
            if res.probe is not None:
                fp = self._safe_fingerprint(res.probe)
                if fp:
                    inventory.record(fp, KeyRecord(profile=h.profile, path=ident))

        result = ImportResult(dry_run=dry_run, keys_found=len(inventory.keys))
        result.adopted = adopted
        for h in parsed:
            result.profiles[h.profile] = result.profiles.get(h.profile, 0) + 1
        if not dry_run:
            manifest.save(self._paths.manifest)
            inventory.save(self._paths.inventory)
        return result

    def _resolve_key(self, h: ParsedHost) -> _Resolution:
        if not h.identity_file:
            # No key on disk to adopt; reconcile will mint a canonical one.
            return _Resolution(derive_key_name(h.profile, h.alias))
        real = Path(os.path.expanduser(h.identity_file))
        pub = real.with_suffix(".pub")
        probe: Path | None = pub if pub.exists() else (real if real.exists() else None)
        if self._under_profile(real, h.profile):
            # Already canonical - keep its filename; nothing to adopt.
            return _Resolution(real.name, adopt_from=None, probe=probe)
        # Non-canonical location (e.g. ~/.ssh/id_ed25519): adopt it into the
        # profiles/ layout under a canonical name so the manifest is faithful.
        canonical = derive_key_name(h.profile, h.alias)
        return _Resolution(
            canonical, adopt_from=real if real.exists() else None, probe=probe
        )

    @staticmethod
    def _under_profile(real: Path, profile: str) -> bool:
        parts = real.parts
        return "profiles" in parts and parts[parts.index("profiles") + 1 :][:1] == (profile,)

    def _adopt_key(self, src_priv: Path, profile: str, key_name: str) -> bool:
        """Copy an existing keypair into profiles/<profile>/<key_name> (perms set).
        Non-destructive: skips if the destination already exists."""
        dst_priv = self._paths.ssh_dir / "profiles" / profile / key_name
        if dst_priv.exists() or not src_priv.exists():
            return False
        dst_priv.parent.mkdir(parents=True, exist_ok=True)
        self._platform.set_perms(dst_priv.parent, perms.DIR_MODE)
        shutil.copy2(src_priv, dst_priv)
        self._platform.set_perms(dst_priv, perms.PRIVATE_KEY_MODE)
        src_pub, dst_pub = src_priv.with_suffix(".pub"), dst_priv.with_suffix(".pub")
        if src_pub.exists():
            shutil.copy2(src_pub, dst_pub)
            self._platform.set_perms(dst_pub, perms.PUBLIC_KEY_MODE)
        return True

    def _safe_fingerprint(self, probe: Path) -> str | None:
        try:
            return self._keystore.fingerprint(probe)
        except Exception:
            return None

File: internal/services/importer/importer.go
// Package importer onboards an existing ~/.ssh into the manifest + inventory,
// ported from services/importer.py. It parses an ssh config (following relative
// Include directives) into profiles + hosts, fingerprints any private keys it can
// find into the inventory, and adopts non-canonical keys into the profiles/ layout.
// Profile assignment derives from an IdentityFile under profiles/<profile>/, else
// "imported".
package importer

import (
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"unicode"

	"github.com/simtabi/ssh-manager/internal/core/inventory"
	"github.com/simtabi/ssh-manager/internal/core/key"
	"github.com/simtabi/ssh-manager/internal/core/manifest"
	"github.com/simtabi/ssh-manager/internal/services/keystore"
	"github.com/simtabi/ssh-manager/internal/util/paths"
	"github.com/simtabi/ssh-manager/internal/util/perms"
)

// kv is one ordered raw_option carried through from the source config.
type kv struct{ Key, Val string }

// ParsedHost is one host block parsed from an ssh config.
type ParsedHost struct {
	Alias        string
	Hostname     string
	User         string
	Port         int
	IdentityFile string // "" if none
	Profile      string
	Extra        []kv
}

// ImportResult summarizes an import run.
type ImportResult struct {
	Profiles  map[string]int // profile -> host count (counts all parsed hosts)
	KeysFound int
	Adopted   int
	DryRun    bool
}

// Format renders the human-readable import summary (mirrors ImportResult.format).
func (r ImportResult) Format() string {
	head := "import:"
	if r.DryRun {
		head = "import (dry-run):"
	}
	lines := []string{head}
	names := make([]string, 0, len(r.Profiles))
	for p := range r.Profiles {
		names = append(names, p)
	}
	sort.Strings(names)
	for _, p := range names {
		lines = append(lines, "  profile "+p+": "+strconv.Itoa(r.Profiles[p])+" host(s)")
	}
	lines = append(lines, "  keys fingerprinted into inventory: "+strconv.Itoa(r.KeysFound))
	if r.Adopted > 0 {
		lines = append(lines, "  keys adopted into profiles/ layout: "+strconv.Itoa(r.Adopted))
	}
	return strings.Join(lines, "\n")
}

var providerByHost = map[string]string{
	"github.com": "github", "gitlab.com": "gitlab", "bitbucket.org": "bitbucket",
}

func inferProvider(hostname string) *string {
	if p, ok := providerByHost[strings.ToLower(hostname)]; ok {
		return &p
	}
	return nil
}

var simpleKeys = map[string]bool{"hostname": true, "user": true, "port": true, "identityfile": true}

// parseSSHConfig parses ssh config text into hosts, following relative Include
// files. Wildcard/global Host blocks are skipped; Match ends the current block.
func parseSSHConfig(text, baseDir string, seen map[string]bool) []*ParsedHost {
	var hosts, current []*ParsedHost
	for _, raw := range strings.Split(text, "\n") {
		line := strings.TrimSpace(raw)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		keyword, value := splitKV(line)
		switch {
		case keyword == "host":
			current = nil
			for _, alias := range strings.Fields(value) {
				if strings.ContainsAny(alias, "*?!") {
					continue
				}
				ph := &ParsedHost{Alias: alias, Port: 22, Profile: "imported"}
				current = append(current, ph)
				hosts = append(hosts, ph)
			}
		case keyword == "match":
			current = nil
		case keyword == "include" && baseDir != "":
			hosts = append(hosts, parseIncludes(value, baseDir, seen)...)
			current = nil
		default:
			for _, ph := range current {
				applyOption(ph, keyword, value)
			}
		}
	}
	return hosts
}

// splitKV splits a config line into a lowercased keyword and the trimmed rest,
// mirroring Python's line.split(None, 1) (whitespace only; no "=" handling).
func splitKV(line string) (string, string) {
	i := strings.IndexFunc(line, unicode.IsSpace)
	if i < 0 {
		return strings.ToLower(line), ""
	}
	return strings.ToLower(line[:i]), strings.TrimSpace(line[i:])
}

func applyOption(h *ParsedHost, keyword, value string) {
	switch keyword {
	case "hostname":
		h.Hostname = value
	case "user":
		h.User = value
	case "port":
		if n, err := strconv.Atoi(value); err == nil {
			h.Port = n
		}
	case "identityfile":
		h.IdentityFile = value
		h.Profile = profileFromIdentity(value)
	default:
		if manifest.IsDangerousOption(keyword) {
			return // drop command/config-executing directives
		}
		if !simpleKeys[keyword] {
			h.Extra = append(h.Extra, kv{keyword, value})
		}
	}
}

func profileFromIdentity(identityFile string) string {
	parts := strings.Split(filepath.ToSlash(identityFile), "/")
	for i, p := range parts {
		if p == "profiles" && i+1 < len(parts) {
			return parts[i+1]
		}
	}
	return "imported"
}

func parseIncludes(pattern, baseDir string, seen map[string]bool) []*ParsedHost {
	var hosts []*ParsedHost
	for _, token := range strings.Fields(pattern) {
		expanded := expanduser(token)
		full := expanded
		if !filepath.IsAbs(expanded) {
			full = filepath.Join(baseDir, expanded)
		}
		matches, _ := filepath.Glob(full)
		sort.Strings(matches)
		for _, match := range matches {
			abs, err := filepath.Abs(match)
			if err != nil {
				continue
			}
			if resolved, e := filepath.EvalSymlinks(abs); e == nil {
				abs = resolved
			}
			if seen[abs] || !isFile(abs) {
				continue
			}
			seen[abs] = true
			b, err := os.ReadFile(abs)
			if err != nil {
				continue
			}
			hosts = append(hosts, parseSSHConfig(string(b), baseDir, seen)...)
		}
	}
	return hosts
}

// resolution is how one parsed host maps onto the canonical layout.
type resolution struct {
	keyName   string
	adoptFrom string // source key to copy into profiles/<p>/ ("" if none)
	probe     string // existing key to fingerprint ("" if none)
}

// Importer onboards an ssh config.
type Importer struct {
	p               paths.Paths
	emitUseKeychain bool
	ks              *keystore.KeyStore
}

// New builds an importer.
func New(p paths.Paths, emitUseKeychain bool) *Importer {
	return &Importer{p: p, emitUseKeychain: emitUseKeychain, ks: keystore.New()}
}

// Run parses configPath into a fresh manifest + inventory (replacing any existing).
// With dryRun it parses and reports without writing or adopting keys.
func (im *Importer) Run(configPath string, dryRun bool) (ImportResult, error) {
	configPath = expanduser(configPath)
	if !isFile(configPath) {
		return ImportResult{}, &importError{"no ssh config file to import: " + configPath}
	}
	text, err := os.ReadFile(configPath)
	if err != nil {
		return ImportResult{}, &importError{"cannot read " + configPath + ": " + err.Error()}
	}
	parsed := parseSSHConfig(string(text), filepath.Dir(configPath), map[string]bool{})

	m := manifest.Starter(im.emitUseKeychain) // defaults with the right global options
	order := []string{}
	byProfile := map[string][]manifest.Host{}
	type rh struct {
		h   *ParsedHost
		res resolution
	}
	var resolved []rh
	seen := map[string]bool{}
	for _, h := range parsed {
		k := h.Profile + "\x00" + h.Alias
		if seen[k] {
			continue // duplicate Host block in the source - first wins
		}
		seen[k] = true
		res := im.resolveKey(h)
		resolved = append(resolved, rh{h, res})
		kn := res.keyName
		hostName := h.Hostname
		if hostName == "" {
			hostName = h.Alias
		}
		user := h.User
		if user == "" {
			if user = os.Getenv("USER"); user == "" {
				user = "git"
			}
		}
		host := manifest.Host{
			Alias: h.Alias, Hostname: hostName, User: user, Port: h.Port,
			Provider: inferProvider(hostName), KeyName: &kn,
			RawOptions: manifest.NewOrderedOptions(pairs(h.Extra)),
		}
		if _, ok := byProfile[h.Profile]; !ok {
			order = append(order, h.Profile)
		}
		byProfile[h.Profile] = append(byProfile[h.Profile], host)
	}
	for _, name := range order {
		m.SetProfile(name, manifest.Profile{KeyScope: "per_service", Hosts: byProfile[name]})
	}

	inv := inventory.New()
	adopted := 0
	for _, r := range resolved {
		ident := m.IdentityFile(r.h.Profile, r.res.keyName)
		if !dryRun && r.res.adoptFrom != "" && im.adoptKey(r.res.adoptFrom, r.h.Profile, r.res.keyName) {
			adopted++
		}
		if r.res.probe != "" {
			if fp := im.safeFingerprint(r.res.probe); fp != "" {
				inv.Record(fp, inventory.KeyRecord{
					Profile: r.h.Profile, Path: ident, Type: "ed25519", RotateAfterDays: 365,
				})
			}
		}
	}

	result := ImportResult{DryRun: dryRun, KeysFound: len(inv.Keys), Adopted: adopted, Profiles: map[string]int{}}
	for _, h := range parsed {
		result.Profiles[h.Profile]++
	}
	if !dryRun {
		if err := m.Save(im.p.Manifest()); err != nil {
			return ImportResult{}, err
		}
		if err := inv.Save(im.p.Inventory()); err != nil {
			return ImportResult{}, err
		}
	}
	return result, nil
}

func (im *Importer) resolveKey(h *ParsedHost) resolution {
	if h.IdentityFile == "" {
		kn, _ := key.DeriveKeyName(h.Profile, h.Alias, "ed25519")
		return resolution{keyName: kn}
	}
	real := expanduser(h.IdentityFile)
	pub := real + ".pub"
	probe := ""
	switch {
	case isFile(pub):
		probe = pub
	case exists(real):
		probe = real
	}
	if underProfile(real, h.Profile) {
		return resolution{keyName: filepath.Base(real), probe: probe}
	}
	kn, _ := key.DeriveKeyName(h.Profile, h.Alias, "ed25519")
	adopt := ""
	if exists(real) {
		adopt = real
	}
	return resolution{keyName: kn, adoptFrom: adopt, probe: probe}
}

func underProfile(real, profile string) bool {
	parts := strings.Split(filepath.ToSlash(real), "/")
	for i, p := range parts {
		if p == "profiles" && i+1 < len(parts) {
			return parts[i+1] == profile
		}
	}
	return false
}

// adoptKey copies an existing keypair into profiles/<profile>/<keyName>.
// Non-destructive: skips if the destination already exists.
func (im *Importer) adoptKey(srcPriv, profile, keyName string) bool {
	dstPriv := filepath.Join(im.p.SSHDir, "profiles", profile, keyName)
	if exists(dstPriv) || !exists(srcPriv) {
		return false
	}
	if err := os.MkdirAll(filepath.Dir(dstPriv), perms.DirMode); err != nil {
		return false
	}
	_ = perms.SetPerms(filepath.Dir(dstPriv), perms.DirMode)
	if !copyFile(srcPriv, dstPriv) {
		return false
	}
	_ = perms.SetPerms(dstPriv, perms.PrivateKeyMode)
	if exists(srcPriv + ".pub") {
		if copyFile(srcPriv+".pub", dstPriv+".pub") {
			_ = perms.SetPerms(dstPriv+".pub", perms.PublicKeyMode)
		}
	}
	return true
}

func (im *Importer) safeFingerprint(probe string) string {
	fp, err := im.ks.Fingerprint(probe)
	if err != nil {
		return ""
	}
	return fp
}

// --- small helpers ---------------------------------------------------------

type importError struct{ msg string }

func (e *importError) Error() string { return e.msg }

func pairs(extra []kv) [][2]string {
	out := make([][2]string, len(extra))
	for i, e := range extra {
		out[i] = [2]string{e.Key, e.Val}
	}
	return out
}

func expanduser(p string) string {
	if p == "~" || strings.HasPrefix(p, "~/") {
		if home, err := os.UserHomeDir(); err == nil {
			if p == "~" {
				return home
			}
			return filepath.Join(home, p[2:])
		}
	}
	return p
}

func exists(path string) bool { _, err := os.Stat(path); return err == nil }

func isFile(path string) bool {
	fi, err := os.Stat(path)
	return err == nil && fi.Mode().IsRegular()
}

func copyFile(src, dst string) bool {
	b, err := os.ReadFile(src)
	if err != nil {
		return false
	}
	return os.WriteFile(dst, b, 0o600) == nil
}
